odbc_fdw.c

```diff
@@ -875,6 +888,205 @@ odbcGetTableSize(odbcFdwOptions* options, unsigned int *size)
+  SQLLEN StrLen_or_Ind;
+} DataBinding;
+
+typedef struct {
+  Oid serverOid;
+  DataBinding* tableResult;
+  SQLHSTMT stmt;
+  SQLCHAR schema;
+  SQLCHAR name;
+  SQLUINTEGER rowLimit;
+  SQLUINTEGER currentRow;
+} TableDataCtx;
+
+
+Datum odbc_tables_list(PG_FUNCTION_ARGS)
+{
+	SQLHENV env;
+	SQLHDBC dbc;
+	SQLHSTMT stmt;
+	SQLRETURN ret;
+  SQLUSMALLINT i;
+  SQLUSMALLINT numColumns = 5;
+  SQLUSMALLINT bufferSize = 1024;
+  SQLUINTEGER rowLimit;
+  SQLUINTEGER currentRow;
+  SQLRETURN retCode;
+
+  FuncCallContext *funcctx;
+  TupleDesc tupdesc;
+  TableDataCtx *datafctx;
+  MemoryContext oldcontext;
+  DataBinding* tableResult;
+  AttInMetadata *attinmeta;
+
+  if (SRF_IS_FIRSTCALL()) {
+    funcctx = SRF_FIRSTCALL_INIT();
+    MemoryContext oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+    datafctx = (TableDataCtx *) palloc(sizeof(TableDataCtx));
+    tableResult = (DataBinding*) palloc( numColumns * sizeof(DataBinding) );
+    
+    char *serverName = text_to_cstring(PG_GETARG_TEXT_PP(0));
+    int serverOid = oid_from_server_name(serverName);
+
+    rowLimit = PG_GETARG_INT32(1); 
+    currentRow = 0;
+
+    odbcFdwOptions options;
+    odbcGetOptions(serverOid, NULL, &options);
+    odbc_connection(&options, &env, &dbc);
+    SQLAllocHandle(SQL_HANDLE_STMT, dbc, &stmt);
+
+    for ( i = 0 ; i < numColumns ; i++ ) {
+      tableResult[i].TargetType = SQL_C_CHAR;
+      tableResult[i].BufferLength = (bufferSize + 1);
+      tableResult[i].TargetValuePtr = palloc( sizeof(char)*tableResult[i].BufferLength );
+    }
+
+    for ( i = 0 ; i < numColumns ; i++ ) {
+      retCode = SQLBindCol(stmt, i + 1, tableResult[i].TargetType, tableResult[i].TargetValuePtr, tableResult[i].BufferLength, &(tableResult[i].StrLen_or_Ind));
+    }
+
+    if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+      ereport(ERROR,
+          (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+           errmsg("function returning record called in context "
+             "that cannot accept type record")));
+
+    attinmeta = TupleDescGetAttInMetadata(tupdesc);
+
+    datafctx->serverOid = serverOid;
+    datafctx->tableResult = tableResult;
+    datafctx->stmt = stmt;
+    datafctx->rowLimit = rowLimit;
+    datafctx->currentRow = currentRow;
+    funcctx->user_fctx = datafctx;
+    funcctx->attinmeta = attinmeta;
+
+    MemoryContextSwitchTo(oldcontext);
+  }
+
+  funcctx = SRF_PERCALL_SETUP();
+
+  datafctx = funcctx->user_fctx; 
+  stmt = datafctx->stmt;
+  tableResult = datafctx->tableResult;
+  rowLimit = datafctx->rowLimit;
+  currentRow = datafctx->currentRow;
+  attinmeta = funcctx->attinmeta;
+
+  retCode = SQLTables( stmt, NULL, SQL_NTS, NULL, SQL_NTS, NULL, SQL_NTS, (SQLCHAR*)"TABLE", SQL_NTS );
+  if (SQL_SUCCEEDED(retCode = SQLFetch(stmt)) && (rowLimit == 0 || currentRow < rowLimit)) {
+    char       **values;
+    HeapTuple    tuple;
+    Datum        result;
+
+    values = (char **) palloc(2 * sizeof(char *));
```

Comment:
[this comment is mostly to help my future self understand what's going on here, but you can read it to look for misunderstandings]

In the first call to the function the `funcctx->multi_call_memory_ctx` memory context was used to allocate data that must survive until the last call (SFR_RETURN_DONE), which seems the right thing to do:

From https://www.postgresql.org/docs/9.5/static/xfunc-c.html

>     /*
     * memory context used for structures that must live for multiple calls
     *
     * multi_call_memory_ctx is set by SRF_FIRSTCALL_INIT() for you, and used
     * by SRF_RETURN_DONE() for cleanup. It is the most appropriate memory
     * context for any memory that is to be reused across multiple calls
     * of the SRF.
     */
    MemoryContext multi_call_memory_ctx;

Now we're allocating the `values` using the default context (good only during the current call), but we should be OK, because the values are copied elsewhere by `BuildTupleFromCStrings` (aren't they?)